### DISCUSSION TITLE
Benchmarking XML readers/writer for pickling

### DISCUSSION THREAD

**Discussion by whophil** [pyvista/pyvista#3279] [Ideas]
## Background
The current approach for pickling PyVista `DataObject`s may have problems on some large datasets (see #3269 - still a live discussion). This has triggered renewed interest in pickling via XML, rather than the "legacy" format currently used

@MatthewFlamm first brought up the idea of using the XML readers/writers for pickling in https://github.com/pyvista/pyvista/pull/1778#issuecomment-959242998.

I took a crack at implementing XML readers/writers for pickling on a branch at https://github.com/whophil/pyvista/tree/feature/xmlpickle. This passes the pickling tests for various dataset types. The switch-like statements exist because the generic XML readers and writers are [broken](https://gitlab.kitware.com/vtk/vtk/-/issues/18661). The "switches" can be removed in the future when those readers/writers are fixed in VTK.

However, in benchmarking, the XML pickling seems to be considerably slower than the legacy format (see results below). **Compression is disabled** in the XML tests. Enabling compression slows down the serialization/de-serialization, as expected.

This leaves us here:
- The legacy format is fast, but fails when attempting to pickle large data objects
- The XML format is slow, but robust

I don't have any suggestions at this point, but I wanted to record this info somewhere.

CC @adeak @akaszynski @banesullivan @adam-grant-hendry 

## Benchmarking results

```python
import pyvista as pv
import pickle
import time

p = [2, 3, 4, 5, 6, 7, 8]
for pi in p:
    c = pv.ParametricRandomHills(randomseed=1, u_res=100 * pi, v_res=100 * pi)
    t0 = time.time()
    pickle.loads(pickle.dumps(c))
    print(c.n_cells, time.time() - t0)
```

produces the following times:

| N points | Legacy [s] | XML [s] |
|----------|------------|---------|
|    40000 |     0.0172 |  0.0637 |
|    90000 |     0.0400 |  0.1213 |
|   160000 |     0.0560 |  0.2164 |
|   250000 |     0.0921 |  0.3425 |
|   360000 |     0.1388 |  0.4771 |
|   490000 |     0.1859 |  0.6662 |
|   640000 |     0.2687 |  0.8520 |

<img width="473" alt="image" src="https://user-images.githubusercontent.com/2186528/188805184-f1f3987f-065a-4069-9070-ec553d831e4f.png">

## Related
- ASCII pickling using the legacy format was implemented in https://github.com/pyvista/pyvista/pull/1143
- Binary pickling using the legacy format was implemented in https://github.com/pyvista/pyvista/pull/1778

**Comment by banesullivan** [comment]
I'm in favor of "slow, but robust" and think we should implement the `__setstate__`/`__getstate__` methods for each datatype to do it's own IO with an appropriate reader/writer

**Reply by MatthewFlamm** [reply]
I agree, that robust is very important here.  It will be impossible to have the ideal code for all situations here.  Some use cases might need more compression, some might need more speed, etc.  Choosing the recommended modern vtk data file standard seems like a sensible approach now that you've figured out the bug in vtk and how to work around it.

**Reply by akaszynski** [reply]
Agreed with the above. Users should have the option for the legacy writer but we should default/recommend the modern format.

**Reply by adeak** [reply]
I concur. Just make sure to point out the incompatibility of newer pyvista versions with older serialization formats. Shouldn't affect the bulk of use cases, but you never know.

**Reply by adam-grant-hendry** [reply]
I'm in favor of not pickling from a security risk standpoint: It's well known that untrusted sources can inject arbitrary byte code into a pickled object that can get run (i.e. [man-in-the-middle attacks](https://realpython.com/python-pickle-module/#security-concerns-with-the-python-pickle-module)). Transferring data between programs or storing data between sessions introduces a security hole. All it would take is one bad apple creating an Issue saying "I can't get my data to load", uploading their pickled file, and a developer trying to load it on their end to help them out and cause havoc (or worse pushing a PR to test a pickled dataset that gets run in our test runners).

I actually originally started out trying to see if I could serialize `VTK` objects using JSON. XML DTD (Document Type Definition) validation and external entity expansion are enabled by default, which makes it open to some attacks unless explicitly disabled. JSON is typically safe, except for JSONP (JSON with Padding), which can result in a CSRF (Cross-Site Request Forgery) attack.

JSON syntax closely resembles python dictionaries, which is nice. In either case, python also has builtin module support for `xml` and `json`.

It was easy enough for me to pickle my VTK objects without `writer.Write()` by simply pulling out the `numpy` point and cell data in my own `__getstate__()` and recreating the objects in `__setstate__()`. I think we could show our end users how to do this in examples, but I think it's safer on our end to only internally support a safer approach like XML or JSON. (We should probably also state we cannot help end users with pickled datasets in Issues and PRs due to these security vulnerabilities).

**Reply by adam-grant-hendry** [reply]
On the pickle end, you can add a cryptographic signature to pickled data with HMAC, but that typically only protects you when sending your own data across a network to ensure it hasn't been tampered with (you can't know if a user is giving you a false HMAC signature ahead of time).

**Reply by adam-grant-hendry** [reply]
w.r.t. JSON vs XML, I think XML is probably the right approach since JSON doesn't have namespaces, so you can't extend or describe JSON values. The tradeoff is XML is complex and can make larger file sizes than JSON.

**Comment by adam-grant-hendry** [comment]
@whophil I see the `pickle` code in [your benchmarks](https://github.com/pyvista/pyvista/discussions/3279#discussion-4369368), but not the `XML` dataset loading/unloading. Perhaps I missed it? Can you show me the source for that benchmark as well?

**Reply by whophil** [reply]
@adam-grant-hendry `pickle.dumps()` and `pickle.loads()` invoke the `__getstate__()` and `__setstate__()` methods of the object being pickled. These methods are responsible for taking the VTK object and converting it to a type that pickle natively knows how to handle - like a `bytes` or `str`.

The code which does this is available in the branch I linked in the original message, and can be found here:
https://github.com/whophil/pyvista/blob/e875dec9c2ac3e308627f4cd409ed5cb9bd25a9d/pyvista/core/dataobject.py#L537-L592

As you can see, it uses the appropriate XML writer to generate an XML-formatted string containing the data of the VTK object, which is later unpacked in `__setstate__()` by the appropriate XML reader.

The "legacy" implementations of `__getstate__()` and `__setstate__()` are what is currently on master, and can be found here:
https://github.com/pyvista/pyvista/blob/d9275790df054fba0682150795f4fdb08266c388/pyvista/core/dataobject.py#L537-L564

These methods do similar to what is done in the XML implementation, except that they use `vtkDataWriter` to generate `bytes`.

**Reply by adam-grant-hendry** [reply]
Thank you sir! Makes sense now.

**Comment by adam-grant-hendry** [comment]
@MatthewFlamm Do you also know if using a parallel XML reader/writer would be appropriate for a large dataset? I haven't used those before, so I don't know.

**Reply by adam-grant-hendry** [reply]
Replying to my own comment here for posterity: parallel readers/writers are for partitioned datasets (e.g. a dataset split across mutliple files, perhaps in an out-of-core dataset situation), so you won't see a read/write speed performance benefit from using them.
